src/text_renderer.rs

```diff
@@ -26,34 +26,32 @@ use gfx::batch::bind;
 
 use gfx::shade::TextureParam;
 
-use gfx_device_gl::{
-    GlDevice,
-    GlResources,
-};
-
 use bitmap_font::BitmapFont;
 use utils::{grow_buffer, MAT4_ID};
 
-pub struct TextRenderer {
-    program: ProgramHandle<GlResources>,
+pub struct TextRenderer<D: Device> {
+    program: ProgramHandle<D::Resources>,
     state: DrawState,
     bitmap_font: BitmapFont,
     vertex_data: Vec<Vertex>,
     index_data: Vec<u32>,
-    vertex_buffer: BufferHandle<GlResources, Vertex>,
-    index_buffer: BufferHandle<GlResources, u32>,
-    params: TextShaderParams<GlResources>,
+    vertex_buffer: BufferHandle<D::Resources, Vertex>,
+    index_buffer: BufferHandle<D::Resources, u32>,
+    params: TextShaderParams<D::Resources>,
 }
 
-impl TextRenderer {
+impl<D: Device> TextRenderer<D> {
 
     pub fn new(
-        graphics: &mut Graphics<GlDevice>,
+        graphics: &mut Graphics<D>,
         frame_size: [u32; 2],
         initial_buffer_size: usize,
         bitmap_font: BitmapFont,
-        font_texture: TextureHandle<GlResources>,
-    ) -> Result<TextRenderer, ProgramError> {
+        font_texture: TextureHandle<D::Resources>,
+    ) -> Result<TextRenderer<D>, ProgramError> {
+
+        // TODO - Currently only have shaders for OpenGL, and no way to pick the right shader
+        // sources ...
 
         let program = match graphics.device.link_program(VERTEX_SRC.clone(), FRAGMENT_SRC.clone()) {
```

Comment:
Is there already a mechanism or a pattern for choosing the proper shader sources?